src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs

```diff
@@ -0,0 +1,180 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    internal static class UnicodeUtility
+    {
+        /// <summary>
+        /// The Unicode replacement character U+FFFD.
+        /// </summary>
+        public const uint ReplacementChar = 0xFFFDU;
+
+        /// <summary>
+        /// Returns the Unicode plane (0 through 16, inclusive) which contains this code point.
+        /// </summary>
+        public static int GetPlane(uint codePoint)
+        {
+            UnicodeDebug.AssertIsValidCodePoint(codePoint);
+
+            return (int)(codePoint >> 16);
+        }
+
+        /// <summary>
+        /// Returns a Unicode scalar value from two code points representing a UTF-16 surrogate pair.
+        /// </summary>
+        public static uint GetScalarFromUtf16SurrogatePair(uint highSurrogateCodePoint, uint lowSurrogateCodePoint)
+        {
+            UnicodeDebug.AssertIsHighSurrogateCodePoint(highSurrogateCodePoint);
+            UnicodeDebug.AssertIsLowSurrogateCodePoint(lowSurrogateCodePoint);
+
+            // This calculation comes from the Unicode specification, Table 3-5.
+            // Need to remove the D800 marker from the high surrogate and the DC00 marker from the low surrogate,
+            // then fix up the "wwww = uuuuu - 1" section of the bit distribution. The code is written as below
+            // to become just two instructions: shl, lea.
+
+            return (highSurrogateCodePoint << 10) + lowSurrogateCodePoint - ((0xD800U << 10) + 0xDC00U - (1 << 16));
+        }
+
+        /// <summary>
+        /// Given a Unicode scalar value, gets the number of UTF-16 code units required to represent this value.
+        /// </summary>
+        public static int GetUtf16SequenceLength(uint value)
```

Comment:
>GetUtf16SequenceLength(uint value) [](start = 26, length = 34)

will you allow make this API work with value like 0xD800 (not well formed surrogate)?

Reply:
nevermind, this is internal class.

---
In reply to: [232750422](https://github.com/dotnet/coreclr/pull/20935#discussion_r232750422) [](ancestors = 232750422)